### PR TITLE
Please Give me the Code Review(코드 리뷰 부탁드립니다.)

### PR DESCRIPTION
**Pull request by postbirds**
## 개요(Overview)
'게시판 만들기' 프로젝트입니다. Thymleaf와 Spring Boot, MySQL(by docker)를 이용해서 만들었습니다. <br>
This is the 'Creating a Bulletin Board' Project. It is created using Thymeleaf and Spring Boot, MySQL DB(by docker)

## 궁금한 점(한글)
- view 부분에서 thymeleaf 적용과 HTML 관련해서 개선 사항
    - form을 이용해서 post 요청을 보내고, a를 이용해서 get을 하는 방식을 주로 취했습니다. 개선 사항이 있을까요?
    - 각 article에 달린 댓글에서 대댓글같은 것을 달고 싶습니다. 이렇게 하기 위해선 댓글 관련 부분이 table로 이루어지면 안 좋을 것 같습니다. 어떻게 하는 게 좋을까요?
    - 이 외에도 타임리프, 자바스크립트 등의 부족에 대해 개선점 편하게 말씀해주세요!
- Controller파트
    - 함수명, 변수명 등이 적절한가요?
    - controller도 테스트 코드를 작성할 수 있을까요? 어떻게 할 수 있나요?
    - service에서 던진 예외들을 어떻게 잡아서 처리하고, null 처리 등이 잘 되었을까요?  
- Service 파트
    - 테스트 코드 작성을 위해 mockito를 사용했지만 제대로 적용했단 생각이 들지 않아요. 테스트 코드 리뷰 부탁드립니다.
    - repository를 가져와서 사용하는 service 부분에서 던진 예외를 어떻게 잡아서 컨트롤러에서 처리해 줄 수 있을까요?
- Repository 파트
   - Spring data jpa를 사용했기 때문에 별 다른 코드는 없습니다. 하지만 이 부분에서도 Test 코드를 작성할 수 있는지 궁금합니다.
- 어노테이션이 autowired되어 있습니다. 이것 말고 다른 방식을 사용해보고 싶은데 제안해주시면 감사하겠습니다. Configuration 클래스를 따로 만들어서 직접 bean을 넣는 메소드를 두는 게 나을까요?

<h4>코드 리뷰 해 주셔서 감사합니다!!</h4>

## Questions(English)
- Improvements related to thymeleaf application and HTML in the view section.
    - I mainly used the form to send a post request and use a to get it. Are there any imporvements?
    - I'd like to add something like a general comment in the comments on each article. To do this, it'd not be a good idea if the comment-related part were made up a table tag. What should I do?
    - In addition to this, please feel free to tell us about improvements in things like ThymeLeap and JavaScript! 
- Controller
    - Are the function names, variable names, etc. appropriate?
    - Can a controller also write test code? How can I do this?
    - How did you catch and handle exceptions thrown by the service, and was null handing done well?
- Service
    - I used mockito to write test code, but I don't think I apply it properly. Please review the test code.
    - How can I catch the exception thrown by the service part that retrieves and uses the repository and handle it in the controller?
- Repository
    - Since I sued Spring data jpa, there is no special code. However, I wonder if it is possible to write test code in this part as well.
    - Annotations are autowired. I'd like to try a different method than this, so I would appreciate your suggestions. Would it be better to create a separate Configuration class and have a method to directly insert the bean?
 
<h4>Thank you for reviewing my code!!</h4>

### CODE REVIEW

File: src/main/java/com/y/java_board/controller/ArticleController.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @GetMapping("/articles")
+    public String articles(Model model){
+        List<Article> articles = articleService.findArticles();
+        model.addAttribute("articles", articles);
+        return "article/articles";
+    }
+
+    @GetMapping("/articles/new")
```
Comment (preethipattabiraman) [inline src/main/java/com/y/java_board/controller/ArticleController.java:38]: Do not use words like new in API. You can use the below:
1. POST -> /articles (body to insert the data)
2. GET all -> /articles
3. GET one article -> /articles/{article-id}

Like this.

File: src/main/java/com/y/java_board/controller/ArticleController.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            model.addAttribute("article",articleOptional.get());
+            model.addAttribute("commentDto", new CommentDto("", "", id));
+            model.addAttribute("comments", comments);
+            return "/article/detail";
+        }
+        // TODO : 해당 id 가 없다는 안내가 필요할 것 같다.
+        return "redirect:/articles";
+    }
+
+    @GetMapping("/articles/delete/{id}")
```
Comment (preethipattabiraman) [inline src/main/java/com/y/java_board/controller/ArticleController.java]: Please do not have verbs in API. If you want to delete an article, go for DELETE request.

File: src/main/java/com/y/java_board/controller/ArticleController.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    public String deleteArticle(@PathVariable("id") long id, Model model){
+        Article article = articleService.findOne(id)
+                .orElseThrow(()-> new IllegalArgumentException("Invalid article Id : " + id));
+        articleService.deleteOne(id);
+        return "redirect:/articles";
+    }
+
+    @GetMapping("/articles/update/{id}")
```
Comment (preethipattabiraman) [inline src/main/java/com/y/java_board/controller/ArticleController.java:76]: Same here too. Use PUT or PATCH.

File: src/main/java/com/y/java_board/controller/ArticleController.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    public String showUpdateForm(@PathVariable long id, Model model){
+        Article article = articleService.findOne(id)
+                .orElseThrow(()-> new IllegalArgumentException("Invalid article Id: "+ id));
+        model.addAttribute("articleDto", new ArticleDto(article.getTitle(),article.getContent(), article.getWriter()));
+        return "/article/update";
+    }
+
+    @PostMapping("/articles/update/{id}")
```
Comment (preethipattabiraman) [inline src/main/java/com/y/java_board/controller/ArticleController.java]: Do not have POST. Use PUT/PATCH Request

Comment (postbirds) [reply src/main/java/com/y/java_board/controller/ArticleController.java]: Thanks to the [link](https://stackoverflow.com/questions/72744349/thymeleaf-thmethod-delete-put-leads-to-request-method-post-not-supported) you provided, I was able to use delete, put, and patch requests on the form as well. I had some inconveniences using only get and post, but I'm very satisfied after using the requests and mapping annotation.

File: src/main/java/com/y/java_board/controller/ArticleController.java (hunk 3 above)
Comment (postbirds) [reply src/main/java/com/y/java_board/controller/ArticleController.java:76]: Solved it the previous one! It was applied not only to the articles related section, but also to the comments section. I'll also make changes to the work done after the current PR!

File: src/main/java/com/y/java_board/controller/ArticleController.java (hunk 2 above)
Comment (postbirds) [reply src/main/java/com/y/java_board/controller/ArticleController.java]: while using the proper mapping method, the api name has also imporved! Now I know that api name should not use verb!

File: src/main/java/com/y/java_board/controller/ArticleController.java (hunk 1 above)
Comment (postbirds) [reply src/main/java/com/y/java_board/controller/ArticleController.java:38]: <h4>Super thank you!</h4> 
It was also applied appropriately to the comment section. These are important customs that are difficult to understand on my own..!

**Pull request merged**